app/uk/gov/hmrc/gform/validation/ValidationService.scala

```diff
@@ -619,32 +572,117 @@ class ComponentsValidator(
+      case After     => isAfterToday(date, offset)
+      case Precisely => isPreciselyToday(date, offset)
+    }
+
   def isAfterToday(date: LocalDate, offset: OffsetDate)(implicit now: Now[LocalDate]): Boolean =
     date.isAfter(now.apply().plusDays(offset.value.toLong))
 
-  def isAfterConcreteDate(date: LocalDate, concreteDay: LocalDate, offset: OffsetDate): Boolean =
-    date.isAfter(concreteDay.plusDays(offset.value.toLong))
+  def isAfterExactConcreteDate(date: LocalDate, concreteDay: ConcreteDate, offset: OffsetDate): Boolean =
+    date.isAfter(exactConcreteDateToLocalDate(concreteDay).plusDays(offset.value.toLong))
 
   def isBeforeToday(date: LocalDate, offset: OffsetDate)(implicit now: Now[LocalDate]): Boolean =
     date.isBefore(now.apply().plusDays(offset.value.toLong))
 
-  def isBeforeConcreteDate(date: LocalDate, concreteDay: LocalDate, offset: OffsetDate): Boolean =
-    date.isBefore(concreteDay.plusDays(offset.value.toLong))
+  def isBeforeExactConcreteDate(date: LocalDate, concreteDay: ConcreteDate, offset: OffsetDate): Boolean =
+    date.isBefore(exactConcreteDateToLocalDate(concreteDay).plusDays(offset.value.toLong))
+
+  def isPreciselyToday(date: LocalDate, offset: OffsetDate)(implicit now: Now[LocalDate]): Boolean =
+    date.isEqual(now.apply().plusDays(offset.value.toLong))
 
-  def validateConcreteDate(concreteDate: ConcreteDate, dateError: GformError): ValidatedLocalDate =
-    Try(LocalDate.of(concreteDate.year, concreteDate.month, concreteDate.day)) match {
-      case Success(date) => Valid(date)
-      case Failure(ex)   => dateError.invalid
+  def isPreciselyExactConcreteDate(date: LocalDate, concreteDay: ConcreteDate, offset: OffsetDate): Boolean =
+    date.isEqual(exactConcreteDateToLocalDate(concreteDay).plusDays(offset.value.toLong))
+
+  def isSameAbstractDate(date: LocalDate, concreteDay: ConcreteDate): Boolean =
+    concreteDay.getNumericParameters
+      .map {
+        case ExactYear(year)   => date.getYear == year
+        case ExactMonth(month) => date.getMonthValue == month
+        case ExactDay(day)     => date.getDayOfMonth == day
+      }
+      .forall(identity)
+
+  def isFirstOrLastDay(date: LocalDate, concreteDay: ConcreteDate): Boolean = concreteDay.day match {
+    case FirstDay => date.getDayOfMonth == 1
+    case LastDay =>
+      exactConcreteDateToLocalDate(ConcreteDate(ExactYear(date.getDayOfYear), ExactMonth(date.getMonthValue), LastDay)).getDayOfMonth == date.getDayOfMonth
+  }
+
+  def isNextOrPreviousYear(date: LocalDate, concreteDay: ConcreteDate): Boolean = {
+    val areMonthAndDayEqual = isSameAbstractDate(date: LocalDate, concreteDay: ConcreteDate)
+    concreteDay.year match {
+      case Next     => date.getYear == getNextYear && areMonthAndDayEqual
+      case Previous => date.getYear == getPreviousYear && areMonthAndDayEqual
+    }
+  }
+
+  def validateConcreteDate(concreteDate: ConcreteDate, dateError: GformError): Validated[GformError, ConcreteDate] = {
```

Comment:
Implementation of this method is prime example where `Validated` data type shine. So let's use it here: https://typelevel.org/cats/datatypes/validated.html